torchtune/models/clip/_position_embeddings.py

```diff
@@ -100,23 +100,314 @@ def __init__(
+                raise ValueError(
+                    f"Loaded local positional embedding has shape {inpt_pos_embed.shape}, "
+                    f"after interpolation. Expected shape {self.local_token_positional_embedding.shape}."
+                )
+
+        # global
+        inpt_global_pos_embed = state_dict.get(
+            prefix + "global_token_positional_embedding"
+        )
+
+        if inpt_global_pos_embed is not None:
+
+            _, _, inpt_n_tokens_per_tile, _ = inpt_global_pos_embed.shape
+
+            # sanity check
+            if math.sqrt(inpt_n_tokens_per_tile - 1) % 1 != 0:
+                raise ValueError(
+                    f"Loaded local positional embedding has shape {inpt_n_tokens_per_tile=}, "
+                    f"which indicates a grid_size that is not squared. This is currently not supported."
+                )
+
+            # instantiated pos emb
+            (
+                tgt_max_num_tiles_x,
+                tgt_max_num_tiles_y,
+                tgt_n_tokens_per_tile,
+                tgt_embed_dim,
+            ) = self.global_token_positional_embedding.shape
+
+            # resize ckpt to match instantiated shape
+            inpt_global_pos_embed = self._resize_global_position_embedding(
+                global_pos_embed=inpt_global_pos_embed,
+                tgt_max_num_tiles=tgt_max_num_tiles_x,
+                tgt_patch_grid_size=int(math.sqrt(tgt_n_tokens_per_tile - 1)),
+            )
+
+            # update state dict
+            state_dict[
+                prefix + "global_token_positional_embedding"
+            ] = inpt_global_pos_embed
+            if (
+                inpt_global_pos_embed.shape
+                != self.global_token_positional_embedding.shape
+            ):
+                raise ValueError(
+                    f"Loaded global positional embedding has shape {inpt_global_pos_embed.shape}, "
+                    f"after interpolation. Expected shape {self.global_token_positional_embedding.shape}."
+                )
+
+    @staticmethod
+    def _resize_local_position_embedding(
+        inpt_pos_embed: torch.Tensor, tgt_patch_grid_size: int
+    ) -> torch.Tensor:
+        """
+        Interpolates the local position embedding for a vision encoder to accommodate
+        a different number of tokens per tile. This is the only dimension that
+        changes during interpolation.
+
+        Args:
+            inpt_pos_embed (torch.Tensor): The position embeddings tensor to be resized. It
+                has shape [n_tokens_per_tile, emb_dim], where the first token is the CLS token
+                and n_tokens_per_tile = patch_grid_size**2 + 1.
+            tgt_patch_grid_size (int): The target size of each patch grid, i.e.,
+                the square root of the number of tokens per tile, excluding the class token.
+
+        Returns:
+            torch.Tensor: The resized position embeddings tensor of shape
+                [tgt_patch_grid_size, dim].
```

Comment:
Is this right? Shouldn't it be `[n_tokens_per_tile, emb_dim]`, but where `n_tokens_per_tile = (tgt_patch_grid_size ** 2) + 1`?